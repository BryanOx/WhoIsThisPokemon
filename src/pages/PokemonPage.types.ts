export interface CorrectAnswer {
  id: number | null;
  correct: boolean;
  message: string[];
}
