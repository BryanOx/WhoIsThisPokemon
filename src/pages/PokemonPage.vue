<template>
  <h1 v-if="!pokemon">Cargando...</h1>
  <div v-else>
    <h1>Â¿Who is this Pokemon?</h1>
    <PokemonPic :pokemonId="pokemon.id" :showPokemon="showPokemon" />
    <PokemonOptions
      @selection="matchPokemon($event)"
      :correctAnswer="correctAnswer"
      :pokemons="pokemonArr"
    />
    <template v-if="showAnswer">
      <h2 v-if="correctAnswer.message.lenght === 1" class="fade-in">
        <b class="correct">{{ correctAnswer.message }}+ "hola"</b>
      </h2>
      <h2 v-else class="fade-in">
        {{ correctAnswer.message[0] }}
        <b class="incorrect">{{ correctAnswer.message[1] }}</b>
        {{ correctAnswer.message[2] }}
        <b class="correct">{{ correctAnswer.message[3] }}</b>
      </h2>
      <div class="buttonContainer" @click="resetGame">
        <h3 class="reset">Restart Game</h3>
      </div>
    </template>
  </div>
</template>

<script lang="ts">
/* eslint-disable @typescript-eslint/no-explicit-any */
import { defineComponent } from "@vue/runtime-core";
import { PokemonPic, PokemonOptions } from "@/components/";
import { getPokemonOptions } from "@/helpers";
import { CorrectAnswer } from './PokemonPage.types';

export default defineComponent({
  components: {
    PokemonPic,
    PokemonOptions
  },
  data() {
    return {
      pokemonArr: [] as any[],
      pokemon: null as any,
      showPokemon: false,
      showAnswer: false,
      correctAnswer: {
        id: null,
        correct: false,
        message: []
      } as CorrectAnswer
    };
  },
  methods: {
    async mixPokemons() {
      this.pokemonArr = await getPokemonOptions();
      const index = Math.floor(Math.random() * 4);
      this.pokemon = this.pokemonArr[index];
    },
    matchPokemon(id: number) {
      this.showPokemon = true;
      this.showAnswer = true;
      this.correctAnswer = {
        id: id,
        correct: id === this.pokemon.id,
        message:
          id === this.pokemon.id
            ? ["That's correct!"]
            : [
                `Ooopss, looks like you choose`,
                `${this.pokemonArr.filter(item => item.id === id)[0].name},`,
                `the correct one was`,
                `${this.pokemon.name}`
              ]
      };
    },
    resetGame() {
      this.pokemonArr = [];
      this.pokemon = null;
      this.showPokemon = false;
      this.showAnswer = false;
      this.correctAnswer = {
        id: null,
        correct: false,
        message: []
      };
      this.mixPokemons();
    }
  },
  mounted() {
    this.mixPokemons();
  }
});
</script>

<style scoped>
b.correct {
  color: rgb(24, 181, 24) !important;
}

b.incorrect {
  color: rgb(201, 31, 31) !important;
}

.buttonContainer {
  display: flex;
  align-self: center;
  justify-content: center;
}

.reset {
  background-color: white;
  padding: 10px 25px;
  cursor: pointer;
  color: red;
  border-radius: 10px;
  box-shadow: 0px 0px 10px 7px rgba(0, 0, 0, 0.2);
}
</style>
