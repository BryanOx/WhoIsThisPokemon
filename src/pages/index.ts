export { default as PokemonPage } from './PokemonPage.vue';
export * from './PokemonPage.types';
