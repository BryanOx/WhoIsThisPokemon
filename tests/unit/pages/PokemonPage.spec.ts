/* eslint-disable @typescript-eslint/no-explicit-any */
import { shallowMount, VueWrapper } from "@vue/test-utils";

import { CorrectAnswer, PokemonPage } from "@/pages";
import { pokemons } from "../mocks";

describe("PokemonPage <pages>", () => {
  let wrapper: VueWrapper<any>;
  let mixPkmArr: jest.SpyInstance;

  beforeEach(() => {
    if (PokemonPage.methods !== undefined) {
      mixPkmArr = jest.spyOn(PokemonPage.methods as any, "mixPokemons");
      wrapper = shallowMount(PokemonPage);
    }
  });

  function mountWithData() {
    return shallowMount(PokemonPage, {
      data() {
        return {
          pokemonArr: pokemons,
          pokemon: pokemons[3],
          showPokemon: false,
          showAnswer: false,
          correctAnswer: {
            id: null,
            correct: null,
            message: []
          } as CorrectAnswer
        };
      }
    });
  }

  it("Should match snapshot", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("Should call mixPokemons on mount", () => {
    expect(mixPkmArr).toHaveBeenCalled();
  });

  it("Should match the snapshot after load pokemons", () => {
    const wrapper = mountWithData();
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("Children components should exist and have correct attributes", () => {
    const wrapper = mountWithData();
    const PokePic = wrapper.find("pokemon-pic-stub");
    const PokeOpt = wrapper.find("pokemon-options-stub");
    expect(PokePic.exists()).toBeTruthy();
    expect(Number(PokePic.attributes("pokemonid"))).toBe(4);
    expect(PokePic.attributes("showPokemon")).toBeFalsy();
    expect(PokeOpt.exists()).toBeTruthy();
    expect(PokeOpt.attributes("pokemons")).toBeTruthy();
  });

  it("Tests in checkAnswer", async () => {
    const wrapper = mountWithData();
    await wrapper.vm.matchPokemon(4);
    expect(wrapper.find("h2").text()).toBe("That's correct!");
    expect(wrapper.vm.showPokemon).toBeTruthy();
    await wrapper.vm.matchPokemon(1);
    expect(wrapper.vm.correctAnswer.message.join(" ")).toBe(
      "Ooopss, looks like you choose bulbasaur, the correct one was charmander"
    );
  });
});
