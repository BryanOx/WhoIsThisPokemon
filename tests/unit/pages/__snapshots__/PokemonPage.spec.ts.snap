// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage <pages> Should match snapshot 1`] = `<h1>Cargando...</h1>`;

exports[`PokemonPage <pages> Should match the snapshot after load pokemons 1`] = `
<div>
  <h1>Â¿Who is this Pokemon?</h1>
  <pokemon-pic-stub pokemonid="4" showpokemon="false"></pokemon-pic-stub>
  <pokemon-options-stub correctanswer="[object Object]" pokemons="[object Object],[object Object],[object Object],[object Object]"></pokemon-options-stub>
  <!--v-if-->
</div>
`;
